archunit/src/main/java/com/tngtech/archunit/library/modules/syntax/ModulesByAnnotationShould.java

```diff
@@ -0,0 +1,109 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.tngtech.archunit.library.modules.syntax;
+
+import java.lang.annotation.Annotation;
+
+import com.tngtech.archunit.PublicAPI;
+import com.tngtech.archunit.base.DescribedPredicate;
+import com.tngtech.archunit.core.domain.Dependency;
+import com.tngtech.archunit.core.domain.JavaClasses;
+import com.tngtech.archunit.core.domain.PackageMatcher;
+import com.tngtech.archunit.lang.ArchRule;
+import com.tngtech.archunit.library.modules.AnnotationDescriptor;
+import com.tngtech.archunit.library.modules.ArchModule;
+
+import static com.tngtech.archunit.PublicAPI.Usage.ACCESS;
+
+@PublicAPI(usage = ACCESS)
+public interface ModulesByAnnotationShould<ANNOTATION extends Annotation> extends ModulesShould<AnnotationDescriptor<ANNOTATION>> {
+
+    /**
+     * Like {@link #respectTheirAllowedDependencies(DescribedPredicate, ModuleDependencyScope)}, but the allowed dependencies will be automatically
+     * derived from the {@link ANNOTATION} property named {@code annotationPropertyName}. This property *must* be of type {@code String[]}
+     * and contain the {@link ArchModule#getName() names} of the {@link ArchModule}s to which access is allowed.
+     * <br><br>
+     * For example, given the annotation
+     * <pre><code>
+     * &#64;interface MyModule {
+     *   String name();
+     *
+     *   String[] allowedDependencies() default {};
+     * }
+     * </code></pre>
+     * and the annotated root classes
+     * <pre><code>
+     * &#64;MyModule(name = "Module One", allowedDependencies = {"Module Two", "Module Three"})
+     * interface ModuleOneDescriptor {}
+     *
+     * &#64;MyModule(name = "Module Two", allowedDependencies = {"Module Three"})
+     * interface ModuleTwoDescriptor {}
+     *
+     * &#64;MyModule(name = "Module Three")
+     * interface ModuleTwoDescriptor {}
+     * </code></pre>
+     * Then the allowed dependencies between the modules would be
+     * <pre><code>
+     * ,----------.
+     * |Module One| ------------.
+     * `----------'              \
+     *      |                    |
+     *      |                    |
+     *      v                    v
+     * ,----------.         ,------------.
+     * |Module Two| ------> |Module Three|
+     * `----------'         `------------'
+     * </code></pre>
+     *
+     * @param annotationPropertyName The name of the property declared within {@link ANNOTATION} that declares allowed dependencies to other {@link ArchModule}s by name
+     * @param dependencyScope        Allows to adjust which {@link Dependency dependencies} are considered relevant by the rule
+     * @return A {@link ArchRule} to be checked against a set of {@link JavaClasses}
+     */
+    @PublicAPI(usage = ACCESS)
+    ModulesByAnnotationRule<ANNOTATION> respectTheirAllowedDependenciesDeclaredIn(String annotationPropertyName, ModuleDependencyScope dependencyScope);
```

Comment:
I wonder whether we should offer this API in a typesafe way:
```suggestion
    ModulesByAnnotationRule<ANNOTATION> respectTheirAllowedDependenciesDeclaredIn(Function<ANNOTATION, String[]> allowedDependencies, ModuleDependencyScope dependencyScope);
```
(`String[]` could also be `Iterable<String>` or something like that...)

If we don't insist on the rule description _"... should respect their allowed dependencies declared in 'allowedDependencies'"_, users of this API could easily replace the magic literal `"allowedDependencies"` with the method reference `AppModule::allowedDependencies`.

Same for `onlyDependOnEachOtherThroughPackagesDeclaredIn`, obviously.

Reply:
I had actually thought the same thing and even already implemented that change, but then as you say, it came down to the rule text. I wanted to see this in the rule text somehow, to be consistent with other places and to also immediately see from the rule text how this is defined. Then I switched it to a `DescribedFunction`. But by the end with all the bloat I felt the convenience was actually gone :man_shrugging: So it came down to rule text vs type safety and I finally thought, that these pure architecture annotations likely don't change often and if, then I would expect some rule could fail and check it anyway :man_shrugging: That was my thought process at least :thinking: Not having it in the rule text at all also didn't feel good to me :thinking:

Reply:
Hehehe, I see your point. 
I had assumed that a custom `AppModule` annotation wouldn't have multiple ambiguous properties, such that the rule description _"modules defined by annotation `@AppModule` should respect their allowed dependencies"_ might be good enough for a rule
```java
        modules()
                .definedByAnnotation(AppModule.class)
                .should().respectTheirAllowedDependenciesDeclaredIn(AppModule::allowedDependencies, // ...
```
so I would have prefered the version that has less degrees of freedom (allowing for usage errors) and simpler code, but it's just a thought. You're definitely right that it doesn't matter once the rule is written, as it will almost never change.

Reply:
I would keep it like this for now for consistency. Because I think a method `respectTheirAllowedDependenciesDeclaredIn` that only adds `respect their allowed dependencies` to the rule text also has some further complications (e.g. automated testing assumes that the method names really match the textual description).
I we see people complaining about this we can still ponder about adding a safer API...